### PR TITLE
CLN: fix mypy errors in pandas\tests\indexes\interval\test_base.py #28926

### PR DESCRIPTION
**Pull request by srcole**
- [ ] xref #28926
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Hi! I based this change on what I saw in #28947. However, when I test this with `mypy pandas/tests/indexes/interval/test_base.py`, I get this error, which I didn't expect:

```
pandas/tests/indexes/interval/test_base.py:16: error: Incompatible types in assignment (expression has type "Optional[Type[IntervalIndex]]", base class "Base" defined the type as "None")
```

I thought the "Optional" that was added would accept the "None." But, additionally, in my local environment, I also can't see any of the errors listed in issue #28926, so maybe I don't have something set up right.

### CODE REVIEW

File: pandas/tests/indexes/interval/test_base.py [hunk 1]
```diff
@@ -12,7 +13,7 @@ class TestBase(Base):
     in test_interval.py or the specific test file (e.g. test_astype.py)
     """
 
-    _holder = IntervalIndex
+    _holder = IntervalIndex # type: Optional[Type[IntervalIndex]]
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/interval/test_base.py]: Mypy considers the initial assignment as the definition of a variable. If you do not explicitly specify the type of the variable, mypy infers the type based on the static type of the value expression see https://mypy.readthedocs.io/en/latest/type_inference_and_annotations.html#type-inference

so to fix the error reported by mypy here, the explicit type declaration needs to be added in the `Base` class

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @srcole for the PR.

**Comment by simonjayhawkins** [comment]
> also can't see any of the errors listed in issue #28926, so maybe I don't have something set up right.

to see the errors, you would need to comment out the ignore_errors in `setup.cfg`

File: pandas/tests/indexes/interval/test_base.py (hunk 1 above)
Comment (srcole) [reply pandas/tests/indexes/interval/test_base.py]: Ah, thanks for the explanation! I've updated the branch to also set the type in the `Base` class.

File: pandas/tests/indexes/interval/test_base.py [hunk 2]
```diff
@@ -12,7 +13,7 @@ class TestBase(Base):
     in test_interval.py or the specific test file (e.g. test_astype.py)
     """
 
-    _holder = IntervalIndex
+    _holder = IntervalIndex  # type: Optional[Type[IntervalIndex]]
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/interval/test_base.py]: you can probably revert this change now.

File: pandas/tests/indexes/interval/test_base.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import numpy as np
 import pytest
+from typing import Optional, Type
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/interval/test_base.py]: same

File: pandas/tests/indexes/common.py [hunk 4]
```diff
@@ -30,7 +31,7 @@
 class Base:
     """ base class for index sub-class tests """
 
-    _holder = None
+    _holder = None  # type: Optional[Type[IntervalIndex]]
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/common.py]: This should be Type[Index] and will fix other errors e.g.
```
pandas\tests\indexes\test_range.py:22: error: Incompatible types in assignment (expression has type "Type[RangeIndex]", base class "Base" defined the type as "None")
pandas\tests\indexes\timedeltas\test_timedelta.py:31: error: Incompatible types in assignment (expression has type "Type[TimedeltaIndex]", base class "Base" defined the type as "None")
pandas\tests\indexes\period\test_period.py:26: error: Incompatible types in assignment (expression has type "Type[PeriodIndex]", base class "Base" defined the type as "None")
pandas\tests\indexes\datetimes\test_datetimelike.py:10: error: Incompatible types in assignment (expression has type "Type[DatetimeIndex]", base class "Base" defined the type as "None")
```

Comment (srcole) [reply pandas/tests/indexes/common.py]: good call. thanks

File: pandas/tests/indexes/interval/test_base.py (hunk 3 above)
Comment (srcole) [reply pandas/tests/indexes/interval/test_base.py]: removed

File: pandas/tests/indexes/interval/test_base.py (hunk 2 above)
Comment (srcole) [reply pandas/tests/indexes/interval/test_base.py]: removed

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
@srcole generally looking good. however ci failing https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=18915

can you sort imports. see https://dev.pandas.io/docs/development/contributing.html#import-formatting

**Comment by srcole** [comment]
I tested it locally, and it seems this change also fixes:

```
pandas\tests\indexes\test_numeric.py
pandas\tests\indexes\test_category.py
```

So I removed them from `setup.cfg`

**Comment by jbrockmendel** [comment]
@srcole can you rebase

File: setup.cfg [hunk 5]
```diff
@@ -148,31 +148,13 @@ ignore_errors=True
...
-ignore_errors=True
-
-[mypy-pandas.tests.indexes.test_numeric]
-ignore_errors=True
-
-[mypy-pandas.tests.indexes.test_range]
-ignore_errors=True
-
-[mypy-pandas.tests.indexes.timedeltas.test_timedelta]
+[mypy-pandas.tests.indexing.test_coercion]
```
Comment (simonjayhawkins) [inline setup.cfg]: @srcole thanks for the updates.

looks like merge master went awry at some point. can you remove this.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]

**Comment by srcole** [comment]
Thanks for catching that @simonjayhawkins. Yup, I guess I missed that when addressing the merge conflict. Anyway, I removed that line.

**Review by simonjayhawkins — APPROVED** [review]
@srcole lgtm @WillAyd

File: pandas/tests/indexes/common.py [hunk 6]
```diff
@@ -30,7 +31,7 @@
 class Base:
     """ base class for index sub-class tests """
 
-    _holder = None
+    _holder = None  # type: Optional[Type[Index]]
```
Comment (WillAyd) [inline pandas/tests/indexes/common.py]: Does this need to be `Optional`?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/common.py (hunk 6 above)
Comment (simonjayhawkins) [reply pandas/tests/indexes/common.py]: IMO we should keep the `Optional` otherwise when we use 3.6 type annotations we'll get
`pandas\tests\indexes\common.py:34: error: Incompatible types in assignment (expression has type "None", variable has type "Type[Index]")`

The Optional should only be excluded where we have added `something = None  # type: ...` for the purpose of type checking.

In this case `_holder` was set to `None` before the type annotation was added.

so when we move to 3.6, we will change to
`_holder: Optional[Type[Index]] = None`
whereas if the initialisation was purely added for the type annotations we would change to just 
`_holder: Type[Index]`

Comment (WillAyd) [reply pandas/tests/indexes/common.py]: Isn't the latter what we actually want here though? OK to change to 36 annotation now on this PR since 35 CI should be dropped within the next few days

Comment (simonjayhawkins) [reply pandas/tests/indexes/common.py]: > Isn't the latter what we actually want here though?

IMO out of scope here

but to answer your question, in general, with  3.6, we could reduce the number of `assert <some_variable> is not None` that are sometimes needed when class attributes are initialised to `None`, but we would need to evaluate on a case by case basis.

> OK to change to 36 annotation now on this PR since 35 CI should be dropped within the next few days

i'm not sure a good reason to block since this is not guaranteed.

**Comment by WillAyd** [comment]
Thanks @srcole !
